### ISSUE TITLE
Unable to create a VMware vSphere cluster with private registry

### ISSUE THREAD

**Issue by suryatejaboorlu** [rancher/rancher#37449]
**Rancher Server Setup**
- Rancher version: 2.6.4
- Installation option (Docker install/Helm Chart): Helm
   - If Helm Chart, Kubernetes Cluster, and version (RKE1, RKE2, k3s, EKS, etc):RKE1 v1.20.5
- Proxy/Cert Details:

**Information about the Cluster**
- Kubernetes version: 1.20.15-rancher1-1
- Cluster Type (Local/Downstream): Downstream
   - If downstream, what type of cluster? (Custom/Imported or specify provider for Hosted/Infrastructure Provider): VMware vSphere

**User Information**
- What is the role of the user logged in? (Admin/Cluster Owner/Cluster Member/Project Owner/Project Member/Custom): Admin

   
**Describe the bug**
Unable to create a VMware vSphere cluster with a private registry after upgrading the Rancher from 2.6.3 to 2.6.4.  
**To Reproduce**
1. Create a Rancher 2.6.4 cluster
2. Create an RKE template with a private registry and add the private registry ca.crt in the cloud-init file.
3. Try to provision the vSphere cluster with the RKE template.

**Result**

**Expected Result**
Rancher should provision the downstream vSphere cluster as it did in 2.6.3 with the same RKE template.

**Logs**
[1] Unable to find image 'docker-opc-ecv-group-cicd-nexus.rp-ocn.apps.ocn.infra.ftgroup/rancher/rancher-agent:v2.6.4' locally docker: Error response from daemon: Head "[https://docker-opc-ecv-group-cicd-nexus.rp-ocn.apps.ocn.infra.ftgroup/v2/rancher/rancher-agent/manifests/v2.6.4":](https://docker-opc-ecv-group-cicd-nexus.rp-ocn.apps.ocn.infra.ftgroup/v2/rancher/rancher-agent/manifests/v2.6.4%22:) no basic auth credentials. See 'docker run --help'. exit status 125 : exit status 1

**Additional context**
The RKE template included private registry details with authentication, but the docker needs the private registry ca.crt file in /etc/ssl/certs/ca-certificates.crt.rancher of the downstream nodes to authenticate the private registry and pull the images. if we copy the ca.crt file manually, the images are pulled and the cluster is provisioning fine. 

SURE-4527

**Comment by phillipsj** [comment]
Adding this issue:

* https://github.com/rancher/rancher/issues/27166

**Comment by jiaqiluo** [comment]
the bug is reproduced in rancher 2.6.4 

Steps:
- run rancher 2.6.4 docker install 
- add a vSphere  node template, in the template put the cloud init config which contains the private registry ca.crt 
- create a node-drive cluster, choose the above node template, enable `private registry` and enter URL, username and password 

Results:
error in rancher logs 
```
2022/04/26 22:55:13 [ERROR] error syncing 'c-mtsh8/m-5mpbj': handler node-controller: Unable to find image '<pri-reg>/rancher/rancher-agent:v2.6.4' locally
docker: Error response from daemon: Head https://<pri-reg>/v2/rancher/rancher-agent/manifests/v2.6.4: no basic auth credentials.
See 'docker run --help'.
exit status 125
: exit status 1, requeuing
```

------

This does not happen in 2.6.3 when following the same steps to create the cluster. 

in 2.6.3 rancher logs has the following line which is missing from 2.6.4 logs 
`2022/04/26 23:26:01 [INFO] [node-controller-rancher-machine] private registry detected, authenticating jiaqi-rancher263-cluster1 to <my private registry>`

**Comment by cbron** [comment]
We decided not to backport this to 2.4.x, as it is EOL. Instead we will add a release note to 2.4.18 about it.

**Comment by jiaqiluo** [comment]
### Root cause
 When rancher provisions an RKE node-driver cluster, one of the steps is to authenticate the machine to a private registry if one is defined in the cluster's spec. A [recent change](https://github.com/rancher/rancher/pull/36578) of migrating the private registry's password to a secret makes Rancher think there is no private registry in the cluster's spec, so it skips authenticating step.  As a result, the error "no basic auth credentials" is returned.


### What was fixed, or what changes have occurred
  The fix is to add the password to the cluster's spec before the authenticating step. The cluster object is used only within the function, so there should be no negative impact on security.

### Areas or cases that should be tested
We need to test the following scenario which is from the issue: 
- add a vSphere node template, in the template put the cloud-init config which contains the private registry ca.crt
- create a node-drive cluster, choose the above node template, enable private registry, and enter the URL, username, and password
- the cluster should be provision successfully 

### What areas could experience regressions?
Besides the above specific scenario, we should also validate that creating a node-driver cluster with a private registry works. 

  
### Are the repro steps accurate/minimal?
yes. 

### Notes to QA 
please reach out to me for a sample of the cloud-init file

**Comment by jiaqiluo** [comment]
the fix can be validated in latest v2.6-head

**Comment by jiaqiluo** [comment]
the fix can be validated in [v2.6.5-rc7](https://github.com/rancher/rancher/releases/tag/v2.6.5-rc7)

**Comment by slickwarren** [comment]
tested on v2.6.5-rc9:

each of the following, when applicable, were configured via cloud init (as this was the original use case)
* normal private registry (no auth) -- pass
* private registry, self signed certs (p0 use case) -- pass
* private registry, insecure -- pass

also tested with cloud provider (out-of-tree) in the following states, with self-signed registry certs:
* rke2, disabled (1.23) -- pass
* rke2, enabled (1.22) -- pass
  * test workload with image from registry and storage attached -- pass
* rke1, enabled (1.22) -- pass
  * test wkld with image from registry and storage attached -- pass
* rke1, disabled (1.23) -- pass
